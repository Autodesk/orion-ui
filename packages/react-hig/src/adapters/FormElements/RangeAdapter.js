/**
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */
import * as HIG from 'hig.web';

import HIGElement from '../../elements/HIGElement';
import * as PropTypes from 'prop-types';
import createComponent from '../createComponent';

class RangeAdapter extends HIGElement {
  constructor(initialProps) {
    super(HIG.Range, initialProps);
  }

  componentDidMount() {
    if (this.props.required) {
      this.hig.required(this.initialProps.required);
    }
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    for (let i = 0; i < updatePayload.length; i += 2) {
      const propKey = updatePayload[i];
      const propValue = updatePayload[i + 1];

      switch (propKey) {
        case 'disabled': {
          propValue ? this.hig.disable() : this.hig.enable();
          break;
        }
        case 'instructions': {
          this.hig.setInstructions(propValue);
          break;
        }
        case 'minValue': {
          this.hig.setMin(propValue);
          break;
        }
        case 'maxValue': {
          this.hig.setMax(propValue);
          break;
        }
        case 'label': {
          this.hig.setLabel(propValue);
          break;
        }
        case 'required': {
          propValue
            ? this.hig.required(propValue)
            : this.hig.noLongerRequired();
          break;
        }
        case 'step': {
          this.hig.setStep(propValue);
          break;
        }
        case 'value': {
          this.hig.setValue(propValue);
          break;
        }
        case 'onBlur': {
          const dispose = this._disposeFunctions.get('onBlurDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set('onBlurDispose', this.hig.onBlur(propValue));
          break;
        }
        case 'onChange':
          const dispose = this._disposeFunctions.get('onChangeDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set('onChangeDispose', this.hig.onChange(propValue));
          break;
        }
        case 'onFocus': {
          const dispose = this._disposeFunctions.get('onFocusDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set('onFocusDispose', this.hig.onFocus(propValue));
          break;
        }
        default: {
          console.log(`RangeAdapter property ${propKey} is unknown`);
        }
      }
    }
  }
}

const RangeComponent = createComponent(RangeAdapter);

RangeComponent.propTypes = {
  disabled: PropTypes.bool,
  instructions: PropTypes.string,
  label: PropTypes.string,
  minValue: PropTypes.number,
  maxValue: PropTypes.number,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  required: PropTypes.string,
  step: PropTypes.number,
  value: PropTypes.number
};

RangeComponent.__docgenInfo = {
  props: {
    disabled: {
      description: 'prevents interaction with the range'
    },
    instructions: {
      description: 'instruction text for the range'
    },
    label: {
      description: 'label for the range'
    },
    minValue: {
      description: 'minimum value for the range'
    },
    maxValue: {
      description: 'maximum value for the range'
    },
    onBlur: {
      description: 'callback for focus lost from the range'
    },
    onChange: {
      description: 'callback for change + focus lost for the range'
    },
    onFocus: {
      description: 'callback for focus event on the range'
    },
    required: {
      description: 'text indicating that this range field is a required field'
    },
    step: {
      description: 'value of each step between min and max.'
    },
    value: {
      description: 'value for the range input'
    },

  }
};

export default RangeComponent;
