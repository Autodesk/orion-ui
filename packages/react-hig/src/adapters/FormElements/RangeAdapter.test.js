/**
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import RangeAdapter from './RangeAdapter';

const inputId = '1234';

describe('<RangeAdapter>', () => {
  function createHigRange(defaults = {}) {
    const higContainer = document.createElement('div');

    const higRange = new HIG.Range({ ...defaults });

    higRange.mount(higContainer);
    mockDataset(higContainer);

    setLabelForInputId(higContainer);

    return { higRange, higContainer };
  }

  function createOrionRange(props) {
    return <RangeAdapter {...props} />;
  }

  function mockDataset(higContainer) {
    higContainer.querySelector('.hig__range__field__range-values').dataset = {
      rangeMin: 0,
      rangeMax: 0
    }

  }

  function setLabelForInputId(higContainer) {
    // to adjust for the randomly generated id
    const label = higContainer.querySelector('label');
    const input = higContainer.querySelector('input');
    label.setAttribute('for', inputId);
    input.setAttribute('id', inputId);
  }

  it('renders a range', () => {
    const defaults = { name: 'mySpecialField', label: 'foo' };

    const { higRange, higContainer } = createHigRange(defaults);
    const container = document.createElement('div');
    const wrapper = mount(createOrionRange(defaults), {
      attachTo: container
    });
    setLabelForInputId(container);

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a range with initial props', () => {
    const defaults = {
      instructions: "Don't just do something, sit there.",
      label: 'Age of your first pet',
      name: 'mySpecialField',
      placeholder: 'How oldu?',
      value: 10
    };

    const { higRange, higContainer } = createHigRange(defaults);
    const container = document.createElement('div');
    const wrapper = mount(createOrionRange(defaults), {
      attachTo: container
    });
    setLabelForInputId(container);

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a range with udpated props', () => {
    const defaults = {
      name: 'mySpecialField', label: 'bar'
    };

    const { higRange, higContainer } = createHigRange(defaults);
    const orionContainer = document.createElement('div');
    const wrapper = mount(createOrionRange(defaults), {
      attachTo: orionContainer
    });

    setLabelForInputId(orionContainer);

    const nextProps = {
      label: 'How old is your pet?',
      required: 'You really must fill this in.',
      minValue: 1,
      maxValue: 20,
      step: 1,
      value: 1
    };
    mockDataset(orionContainer);

    higRange.setLabel(nextProps.label);
    higRange.required(nextProps.required);
    higRange.setMax(nextProps.maxValue);
    higRange.setMin(nextProps.minValue);
    higRange.setStep(nextProps.step);
    higRange.setValue(nextProps.value);

    wrapper.setProps(nextProps);

    expect(orionContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(orionContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  ['onBlur','onFocus', 'onChange'].forEach(eventName => {
    it(`sets event listeners for ${eventName} initially`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(createOrionRange({ [eventName]: spy }), {
        attachTo: container
      });
      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(eventName+'Dispose');
      expect(disposeFunction).toBeDefined();

    });

    it(`sets event listeners for ${eventName} when updated`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(createOrionRange({}), {
        attachTo: container
      });
      wrapper.setProps({ [eventName]: spy });

      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(eventName+'Dispose');
      expect(disposeFunction).toBeDefined();
    });
  });

});
