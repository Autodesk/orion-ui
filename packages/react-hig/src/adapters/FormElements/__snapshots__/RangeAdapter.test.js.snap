// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeAdapter> renders a text field 1`] = `
"<div class=\\"hig__range \\">
    <div class=\\"hig__range__field__range-values\\" data-range-min=\\"data-range-max=\\">
      <span class=\\"hig__range__field__current-value\\"></span>
      <input id=\\"1234\\" class=\\"hig__range__field\\" type=\\"range\\" min=\\"max=\\" step=\\"1\\">
      <!--LABEL-->
      <label for=\\"1234\\" class=\\"hig__range__label\\">foo</label>
    </div>

    <!--INSTRUCTIONS-->
    
    <!--REQUIRED-NOTICE-->
</div>"
`;

exports[`<RangeAdapter> renders a text field with initial props 1`] = `
"<div class=\\"hig__range \\">
    <div class=\\"hig__range__field__range-values\\" data-range-min=\\"data-range-max=\\">
      <span class=\\"hig__range__field__current-value\\">10</span>
      <input id=\\"1234\\" class=\\"hig__range__field\\" type=\\"range\\" min=\\"max=\\" step=\\"1\\">
      <!--LABEL-->
      <label for=\\"1234\\" class=\\"hig__range__label\\">Age of your first pet</label>
    </div>

    <!--INSTRUCTIONS-->
    <p class=\\"hig__range__instructions\\">Don't just do something, sit there.</p>
    <!--REQUIRED-NOTICE-->
</div>"
`;

exports[`<RangeAdapter> renders a text field with udpated props 1`] = `
"<div class=\\"hig__range hig__range--required\\">
    <div class=\\"hig__range__field__range-values\\" data-range-min=\\"data-range-max=\\">
      <span class=\\"hig__range__field__current-value\\">1</span>
      <input id=\\"1234\\" class=\\"hig__range__field\\" type=\\"range\\" min=\\"1\\" step=\\"1\\" max=\\"20\\" value=\\"1\\">
      <!--LABEL-->
      <label for=\\"1234\\" class=\\"hig__range__label\\">How old is your pet?</label>
    </div>

    <!--INSTRUCTIONS-->
    
    <p class=\\"hig__range__required-notice\\">You really must fill this in.</p><!--REQUIRED-NOTICE-->
</div>"
`;
